test/test_basic_submission.py

```diff
@@ -109,36 +110,47 @@ def test_dockstore_import_in_terra(self):
-
-            print(f"md5sum workflow state is: {response['workflows'][0]['status']}.  Checking again in 20 seconds.")
-            if timeout < 0:
-                print(json.dumps(response, indent=4))
-                raise RuntimeError('The md5sum workflow run timed out.  '
-                                   f'Expected 4 minutes, but took longer than '
-                                   f'{float(twenty_minutes - timeout) / 60.0} minutes.')
-
-        with self.subTest('Dockstore Workflow Run Completed Successfully'):
-            if response['workflows'][0]['status'] != "Succeeded":
-                raise RuntimeError(f'The md5sum workflow did not succeed:\n{json.dumps(response, indent=4)}')
+            with self.subTest('Dockstore Workflow Run Submitted'):
+                self.assertEqual(status, 'Submitted')
+            with self.subTest('Dockstore Workflow Run Responds with DRS.'):
+                self.assertTrue(response['workflows'][0]['inputResolutions'][0]['value'].startswith('drs://'))
+
+            submission_id = response['submissionId']
+
+            # md5sum should run for about 4 minutes, but may take far longer(?); give a generous timeout
+            # also configurable manually via MD5SUM_TEST_TIMEOUT if held in a pending state
+            start = time.time()
+            deadline = start + int(os.environ.get('MD5SUM_TEST_TIMEOUT', 60 * 60))
+            while True:
+                response = check_workflow_status(submission_id=submission_id)
+                status = response['status']
+                if status == 'Done':
+                    break
+                elif response['workflows'][0]['status'] == "Failed":
+                    raise RuntimeError(f'The md5sum workflow did not succeed:\n{json.dumps(response, indent=4)}')
+                else:
+                    now = time.time()
+                    if now < deadline:
+                        print(f"md5sum workflow state is: {response['workflows'][0]['status']}. "
+                              f"Checking again in 20 seconds.")
+                        time.sleep(20)
+                    else:
+                        print(json.dumps(response, indent=4))
+                        raise RuntimeError('The md5sum workflow run timed out.  '
+                                           f'Expected 4 minutes, but took longer than '
+                                           f'{float(start - now) / 60.0} minutes.')
+
+            with self.subTest('Dockstore Workflow Run Completed Successfully'):
+                if response['workflows'][0]['status'] != "Succeeded":
+                    raise RuntimeError(f'The md5sum workflow did not succeed:\n{json.dumps(response, indent=4)}')
+        finally:
+            table = f'platform-dev-178517.bdc.terra_md5_latency_{STAGE}'
+            client = Client()
+            client.add_row(table, time.time() - start)
```

Comment:
This should probably have it's own function to add the row with a retry decorator on it.

There should probably be an extra try except here too when calling that function, since if it does not post the row after multiple retries (because BigQuery is down for whatever reason), but the test passed, we should still report the test success/failure and not fail on the latency post.

Reply:
> This should probably have its own function to add the row with a retry decorator on it.

:+1:

> There should probably be an extra try except here too when calling that function

I considered that, but then we wouldn't know if the latency failed to report.

Reply:
I'll add a try except.